src/ScalarMessaging.js

```diff
@@ -355,7 +379,15 @@ function setWidget(event, roomId) {
             };
         }
 
+        // This starts listening for when the echo comes back from the server
+        // since the widget won't appear added until this happens. If we don't
+        // wait for this, the action will complete but if the user is fast enough,
+        // the widget still won't actually be there.
+        // start listening now otherwise we could race
+        const widgetAddPromise = waitForUserWidget(widgetId);
         client.setAccountData('m.widgets', userWidgets).then(() => {
+            return widgetAddPromise;
+        }).then(() => {
```

Comment:
There's no `catch` here despite us potentially `reject`ing in `widgetAddPromise`.

We should also probably not send "success" or dispatch "user_widget_updated" if we failed to add the widget.

Reply:
This won't send success or do the dispatch if one of those things failed since the 'then' block won't run. The error handling should be handled in the same way as if the `setAccountData` failed.

Reply:
>This won't send success or do the dispatch if one of those things failed 

True, I am blind.

>in the same way as if the setAccountData failed.

At the moment, this isn't handled either.

We need to handle the case where the timeout fires an we reject with an error, otherwise this could propagate as unhandled?